03-rnaseq/pathway-analysis_rnaseq_02_gsea.Rmd

```diff
@@ -0,0 +1,542 @@
+See the [`DESeq2` package vignette](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html) for more information on how `DESeq2` handles the log2 fold change values with the `lfcShrink()` function.
+
+Let's check to see that we removed the duplicate gene symbols and kept the rows with the higher absolute value of the log2 fold change.
+
+```{r}
+filtered_dge_mapped_df %>%
+  dplyr::filter(gene_symbol %in% dup_gene_symbols) %>%
+  dplyr::arrange(gene_symbol)
+```
+
+Looks like we were able to successfully get rid of the duplicate gene identifiers and keep the observations with the higher absolute value of the log2 fold change!
+
+## Perform gene set enrichment analysis (GSEA)
+
+The goal of GSEA is to detect situations where all genes in a gene set change in a coordinated way, detecting even small statistical but coordinated changes.
+This is done by ranking genes within a gene set from most highly positive to most highly negative, weighting them according to their gene-level statistic, to calculate the enrichment score (ES), which is a pathway-level statistic [@clusterProfiler-book].
+
+### Determine our pre-ranked genes list
+
+The `GSEA()` function takes a pre-ranked and sorted named vector of statistics, where the names in the vector are gene identifiers.
+We will therefore need to create a gene list with statistics that GSEA will rank by.
+
+In the next chunk, we will create a named vector ranked based on the gene-level log2 fold change values.
+
+```{r}
+# Let's create a named vector ranked based on the log2 fold change values
+lfc_vector <- filtered_dge_mapped_df$log2FoldChange
+names(lfc_vector) <- filtered_dge_mapped_df$gene_symbol
+
+# We need to sort the log2 fold change values in descending order here
+lfc_vector <- sort(lfc_vector, decreasing = TRUE)
+```
+
+Let's preview our pre-ranked named vector.
+
+```{r}
+# Look at first entries of the ranked log2 fold change vector
+head(lfc_vector)
+```
+
+### Run GSEA using the `GSEA()` function
+
+Genes were ranked from most highly positive to most highly negative, weighted according to their gene-level statistic, in the previous section.
+In this section, we will implement GSEA to calculate the enrichment score (ES) for each gene set using our pre-ranked gene list.
+
+We can use the `GSEA()` function to perform GSEA with any generic set of gene sets, but there are several functions for using specific, commonly used gene sets (e.g., `gseKEGG()`).
+
+```{r}
+gsea_results <- GSEA(
+  geneList = lfc_vector, # ordered ranked gene list
+  minGSSize = 25, # minimum gene set size
+  maxGSSize = 500, # maximum gene set set
+  pvalueCutoff = 0.05, # p value cutoff
+  eps = 0, # boundary for calculating the p value
+  seed = TRUE, # set seed to make results reproducible
+  pAdjustMethod = "BH", # Benjamini-Hochberg correction
+  TERM2GENE = dplyr::select(
+    hs_gene_sets,
+    gs_name,
+    gene_symbol
+  )
+)
+```
+
```

Comment:
> It says there's ties in rankings: `There are ties in the pre-ranked stats (0.01% of the list).` which is weird.
> 
> And what's more weird and doesn't explain anything is that
> `length(unique(filtered_dge_mapped_df$log2FoldChange))` is three numbers shorter than `filtered_dge_mapped_df$log2FoldChange`

I may have misunderstood this comment, but my interpretation is that this means that there are 3 instances of ties which are the three values "missing" from the one that includes `unique()`.

> 
> and still not explaining anything:
> 
> ```
> dups <- which(duplicated(filtered_dge_mapped_df$log2FoldChange))
> filtered_dge_mapped_df$log2FoldChange[dups]
> ```
> 
> This prints out three seemingly distinct numbers. I'm confused.

Quoting the docs from `duplicated()`:

> `duplicated()` determines which elements of a vector or data frame are duplicates of elements with smaller subscripts, and returns a logical vector indicating which elements (rows) are duplicates.

So we would expect different values to be returned by that code block, because it will return the _second_ (or greater) instance of duplicated, or tied, values. Taken with the information from above, I anticipate these are the three values that show up for more than one gene that `GSEA()` is warning about.

Reply:
Yes. Okay. Right. `duplicated()` makes sense, I knew I was misunderstanding something incorrectly, but my brain got lost yesterday. So there's dups. :+1: